build-system/tasks/visual-diff.js

```diff
@@ -35,10 +100,463 @@ function setPercyBranch() {
+
+    await enableExperiments(page, webpage['experiments']);
+    log('verbose', 'Navigating to page', colors.yellow(`${BASE_URL}/${url}`));
+    await page.goto(`${BASE_URL}/${url}`);
+
+    await verifyCssElements(page, url, webpage.forbidden_css,
+        webpage.loading_incomplete_css, webpage.loading_complete_css);
+    await percy.snapshot(name, page, SNAPSHOT_OPTIONS);
+    await clearExperiments(page);
+  }
+}
+
+/**
+ * Verifies that all CSS elements are as expected before taking a snapshot.
+ *
+ * @param {!puppeteer.Page} page a Puppeteer control browser tab/page.
+ * @param {string} url URL to be snapshotted.
+ * @param {!Array<string>} forbiddenCss Array of CSS elements that must not be
+ *     found in the page.
+ * @param {!Array<string>} loadingIncompleteCss Array of CSS elements that must
+ *     eventually be removed from the page.
+ * @param {!Array<string>} loadingCompleteCss Array of CSS elements that must
+ *     eventually appear on the page.
+ */
+async function verifyCssElements(page, url, forbiddenCss, loadingIncompleteCss,
+  loadingCompleteCss) {
+  // Wait for loader dot to be hidden.
+  await page.waitForSelector('.i-amphtml-loader-dot', {
+    hidden: true,
+    timeout: CSS_SELECTOR_TIMEOUT_MS,
+  }).catch(() => {
+    log('fatal', colors.cyan(url),
+        `still has the AMP loader dot after ${CSS_SELECTOR_TIMEOUT_MS} ms`);
+  });
+
+  if (forbiddenCss) {
+    for (const css of forbiddenCss) {
+      if (await page.$(css) !== null) {
+        log('fatal', colors.cyan(url), 'has forbidden CSS element',
+            colors.cyan(css));
+      }
+    }
+  }
+
+  if (loadingIncompleteCss) {
+    for (const css of loadingIncompleteCss) {
+      log('verbose', `Waiting for invisibility of ${css}`);
+      await page.waitForSelector(css, {
+        hidden: true,
+        timeout: CSS_SELECTOR_TIMEOUT_MS,
+      }).catch(() => {
+        log('fatal', colors.cyan(url), 'still has CSS element',
+            colors.cyan(css), `after ${CSS_SELECTOR_TIMEOUT_MS} ms`);
+      });
+    }
+  }
+
+  if (loadingCompleteCss) {
+    for (const css of loadingCompleteCss) {
+      log('verbose', 'Waiting for visibility of', css);
+      await page.waitForSelector(css, {
+        visible: true,
+        timeout: CSS_SELECTOR_TIMEOUT_MS,
+      }).catch(() => {
+        log('fatal', colors.cyan(url), 'does not yet have CSS element',
+            colors.cyan(css), `after ${CSS_SELECTOR_TIMEOUT_MS} ms`);
+      });
+    }
+  }
+}
+
+/**
+ * Enables the given AMP experiments.
+ *
+ * @param {!puppeteer.Page} page a Puppeteer control browser tab/page.
+ * @param {*} experiments List of experiments to enable.
+ */
+async function enableExperiments(page, experiments) {
+  if (experiments) {
+    log('verbose', 'Setting AMP experiments',
+        colors.cyan(experiments.join(', ')));
+    await page.setCookie({
+      name: 'AMP_EXP',
+      value: experiments.join('%2C'),
+      domain: HOST,
+    });
```

Comment:
When I wrote this in Ruby, I had to do some local testing (by printing out the contents of all local cookies) before and after each snapshot to get the code to work correctly. I assume you were able to verify that the JS calls to `page.setCookie` and `page.deleteCookie` work correctly as written?

Reply:
Ack; we checked this together and we're satisfied that this looks correct